[c10d][fr] Fix another bug when we should continue when the op list is empty

fduwjj opened the pull request:
Differential Revision: D73375318

We shouldn't check the op list when it is empty. And later, when it is empty we pops it out from the queue we will check for collective matching. Added a unit test for this case and also covered the case fixed https://github.com/pytorch/pytorch/pull/151683 in the unit test as well.


cc @H-Huang @awgu @wanchaol @fegin @wz337 @wconstab @d4l3k

d4l3k reviewed (approved):
Can we catch these issues with E2E tests? 

Seems like there's a lot of edge cases popping up here and having an E2E test to automatically detect and prevent regressions seems valuable

It may also be useful to save a few small FR traces in a repo so we can use them for backtesting

fduwjj commented:
Thanks for the feedback. We found this bug in the real e2e test indeed and I agree ideally we should keep some for backtesting. 

But there are some challenges there because these dumps are all from Meta internal workloads, lots of cleaning and pruning and non-technical work (legal review) will be needed and we don't have OSS generated FR traces. That's why we tried to mimic these examples and trying to represent these cases in a synthetic way and test them. The current synthetic test is the closest we can go for an e2e test before we could have a full e2e test.

And the reason why we had lots of fix recently is because we didn't include some real cases into the synthetic cases and we are fixing it proactively before users complain. (And the volume of these errors is really low, single digit). I am sure with these new synthetic cases added, later changes will be tested more thoroughly.

Commit:
[c10d][fr] Fix another bug when we should continue when the op list is empty (#151798)

Summary:


We shouldn't check the op list when it is empty. And later, when it is empty we pops it out from the queue we will check for collective matching. Added a unit test for this case and also covered the case fixed https://github.com/pytorch/pytorch/pull/151683 in the unit test as well.

Test Plan: CI

Reviewed By: d4l3k, VieEeEw

Differential Revision: D73375318

wconstab reviewed (approved):
unblock

kwen2501 reviewed (changes requested):
PR description?

fduwjj commented:
@kwen2501 somehow the description was not synced from diff to PR. I just manually copied it.

fduwjj commented:
@pytorchbot merge -f "The request change is addressed"
